flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/ShardConsumer.java

```diff
@@ -379,49 +206,18 @@ private void deserializeRecordForCollectionAndUpdateState(UserRecord record)
 	}
 
 	/**
-	 * Calls {@link KinesisProxyInterface#getRecords(String, int)}, while also handling unexpected
-	 * AWS {@link ExpiredIteratorException}s to assure that we get results and don't just fail on
-	 * such occasions. The returned shard iterator within the successful {@link GetRecordsResult} should
-	 * be used for the next call to this method.
+	 * Filters out aggregated records that have previously been processed.
+	 * This method is to support restarting from a partially consumed aggregated sequence number.
 	 *
-	 * <p>Note: it is important that this method is not called again before all the records from the last result have been
-	 * fully collected with {@link ShardConsumer#deserializeRecordForCollectionAndUpdateState(UserRecord)}, otherwise
-	 * {@link ShardConsumer#lastSequenceNum} may refer to a sub-record in the middle of an aggregated record, leading to
-	 * incorrect shard iteration if the iterator had to be refreshed.
-	 *
-	 * @param shardItr shard iterator to use
-	 * @param maxNumberOfRecords the maximum number of records to fetch for this getRecords attempt
-	 * @return get records result
-	 * @throws InterruptedException
+	 * @param record the record to filter
+	 * @return {@code true} if the record should be retained
 	 */
-	private GetRecordsResult getRecords(String shardItr, int maxNumberOfRecords) throws Exception {
-		GetRecordsResult getRecordsResult = null;
-		while (getRecordsResult == null) {
-			try {
-				getRecordsResult = kinesis.getRecords(shardItr, maxNumberOfRecords);
-
-				// Update millis behind latest so it gets reported by the millisBehindLatest gauge
-				Long millisBehindLatest = getRecordsResult.getMillisBehindLatest();
-				if (millisBehindLatest != null) {
-					shardMetricsReporter.setMillisBehindLatest(millisBehindLatest);
-				}
-			} catch (ExpiredIteratorException eiEx) {
-				LOG.warn("Encountered an unexpected expired iterator {} for shard {};" +
-					" refreshing the iterator ...", shardItr, subscribedShard);
-
-				shardItr = getShardIterator(lastSequenceNum);
-
-				// sleep for the fetch interval before the next getRecords attempt with the refreshed iterator
-				if (fetchIntervalMillis != 0) {
-					Thread.sleep(fetchIntervalMillis);
-				}
-			}
+	private boolean filterDeaggregatedRecord(final UserRecord record) {
+		if (lastSequenceNum.isAggregated()) {
+			return !record.getSequenceNumber().equals(lastSequenceNum.getSequenceNumber()) ||
+				record.getSubSequenceNumber() > lastSequenceNum.getSubSequenceNumber();
 		}
-		return getRecordsResult;
-	}
 
-	@SuppressWarnings("unchecked")
-	protected static List<UserRecord> deaggregateRecords(List<Record> records, String startingHashKey, String endingHashKey) {
-		return UserRecord.deaggregate(records, new BigInteger(startingHashKey), new BigInteger(endingHashKey));
+		return true;
```

Comment:
Just for easier code readability, can we rewrite this as:

```
if (!lastSequenceNum.isAggregated()) {
    return true;
}

return !record.getSequenceNumber().equals(lastSequenceNum.getSequenceNumber()) ||
				record.getSubSequenceNumber() > lastSequenceNum.getSubSequenceNumber();
```

I've found early returns always helps with understanding these :)

Reply:
Done